{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Schema iterative evals\n",
    "In this notebook I'll perform heuristic evaluations of the output of OntoGPT as I iterate over the schema."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/home/lotrecks/anaconda3/envs/ontogpt/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "from ontogpt.io.csv_wrapper import parse_yaml_predictions\n",
    "import pandas as pd\n",
    "import jsonlines\n",
    "from os import listdir\n",
    "from collections import defaultdict\n",
    "import regex"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in data\n",
    "### Gold standard annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "with jsonlines.open('../data/gold_standards/pickle_all.jsonl') as reader:\n",
    "    pickle = [obj for obj in reader]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get only the docs that we applied OntoGPT on\n",
    "evaled_ids = [fname.split('.')[0] for fname in listdir('../data/ontogpt_input/pickle_5/')]\n",
    "gold = [doc for doc in pickle if doc['doc_key'] in evaled_ids]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### OntoGPT output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "to_eval_1 = '../data/ontogpt_output/pickle_5/output.txt'\n",
    "to_eval_2 = '../data/ontogpt_output/pickle_5_iter2/output.txt'\n",
    "to_eval_3 = '../data/ontogpt_output/pickle_5_iter3/output.txt'\n",
    "schema_path = '../knowledge_graph/schema/desiccation.yaml'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Format OntoGPT output as CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 5/5 [00:00<00:00, 17985.87it/s]\n",
      "100%|██████████| 5/5 [00:00<00:00, 17008.53it/s]\n",
      "100%|██████████| 5/5 [00:00<00:00, 24105.20it/s]\n"
     ]
    }
   ],
   "source": [
    "ent_df_1, rel_df_1 = parse_yaml_predictions(to_eval_1, schema_path)\n",
    "ent_df_2, rel_df_2 = parse_yaml_predictions(to_eval_2, schema_path)\n",
    "ent_df_3, rel_df_3 = parse_yaml_predictions(to_eval_3, schema_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>category</th>\n",
       "      <th>name</th>\n",
       "      <th>provided_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0004707</td>\n",
       "      <td>genes</td>\n",
       "      <td>MAPK</td>\n",
       "      <td>765bee80-fe70-4a64-998c-16412276b343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AUTO:SIPK</td>\n",
       "      <td>genes</td>\n",
       "      <td>SIPK</td>\n",
       "      <td>765bee80-fe70-4a64-998c-16412276b343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AUTO:WIPK</td>\n",
       "      <td>genes</td>\n",
       "      <td>WIPK</td>\n",
       "      <td>765bee80-fe70-4a64-998c-16412276b343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AUTO:NahG</td>\n",
       "      <td>genes</td>\n",
       "      <td>NahG</td>\n",
       "      <td>765bee80-fe70-4a64-998c-16412276b343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AUTO:salicylic%20acid-induced%20protein%20kinase</td>\n",
       "      <td>proteins</td>\n",
       "      <td>salicylic acid-induced protein kinase</td>\n",
       "      <td>765bee80-fe70-4a64-998c-16412276b343</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 id  category  \\\n",
       "0                                        GO:0004707     genes   \n",
       "1                                         AUTO:SIPK     genes   \n",
       "2                                         AUTO:WIPK     genes   \n",
       "3                                         AUTO:NahG     genes   \n",
       "4  AUTO:salicylic%20acid-induced%20protein%20kinase  proteins   \n",
       "\n",
       "                                    name                           provided_by  \n",
       "0                                   MAPK  765bee80-fe70-4a64-998c-16412276b343  \n",
       "1                                   SIPK  765bee80-fe70-4a64-998c-16412276b343  \n",
       "2                                   WIPK  765bee80-fe70-4a64-998c-16412276b343  \n",
       "3                                   NahG  765bee80-fe70-4a64-998c-16412276b343  \n",
       "4  salicylic acid-induced protein kinase  765bee80-fe70-4a64-998c-16412276b343  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ent_df_3.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We need to map the original filenames to the ID's assigned by OntoGPT; luckily it seems like the order is preserved when random ID's are assigned."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_doc_keys(ent_df, rel_df, input_dir):\n",
    "    id_map = {rid: orig_id.split('.')[0] for rid, orig_id in zip(ent_df.provided_by.unique(), listdir(input_dir))}\n",
    "    ent_df['doc_key'] = ent_df['provided_by'].map(id_map)\n",
    "    rel_df['doc_key'] = rel_df['provided_by'].map(id_map)\n",
    "    return ent_df, rel_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "ent_df_1, rel_df_1 = map_doc_keys(ent_df_1, rel_df_1, '../data/ontogpt_input/pickle_5/')\n",
    "ent_df_2, rel_df_2 = map_doc_keys(ent_df_2, rel_df_2, '../data/ontogpt_input/pickle_5/')\n",
    "ent_df_3, rel_df_3 = map_doc_keys(ent_df_3, rel_df_3, '../data/ontogpt_input/pickle_5/')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluate entities\n",
    "We'll look at the predictions and the true entities side by side, and assign a manual heuristic score for each abstract as a way to track improvement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "types_to_keep = [\n",
    "    'Amino_acid_monomer',\n",
    "    'Peptide',\n",
    "    'Protein',\n",
    "    'Nucleotide',\n",
    "    'Polynucleotide',\n",
    "    'DNA',\n",
    "    'RNA',\n",
    "    'Organic_compound_other',\n",
    "    'Inorganic_compound_other',\n",
    "    'Element',\n",
    "    'Multicellular_organism',\n",
    "    'Unicellular_organism',\n",
    "    'Virus'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "target_entities = defaultdict(list)\n",
    "for doc in gold:\n",
    "    doc_key = doc['doc_key']\n",
    "    sent_toks = [tok for sent in doc['sentences'] for tok in sent]\n",
    "    for sent in doc['ner']:\n",
    "        for ent in sent:\n",
    "            if ent[2] in types_to_keep:\n",
    "                ent_text = ' '.join(sent_toks[ent[0]:ent[1]+1])\n",
    "                target_entities[doc_key].append((ent_text, ent[2]))\n",
    "target_entities = {k: list(set(v)) for k, v in target_entities.items()}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There is [a bug](https://github.com/monarch-initiative/ontogpt/issues/351) in OntoGPT that puts entities from previous documents in all following documents; let's quickly pre-processes here to remove the suplicates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_preds(evaled_ids, ent_df):\n",
    "    processed_preds = {}\n",
    "    for i, doc in enumerate(evaled_ids):\n",
    "        if i == 0:\n",
    "            processed_preds[doc] = ent_df[ent_df['doc_key'] == doc].name.tolist()\n",
    "        else:\n",
    "            current_list = ent_df[ent_df['doc_key'] == doc].name.tolist()\n",
    "            previous_ents = [\n",
    "                ent for prev_doc in evaled_ids[:i] for ent in processed_preds[prev_doc]\n",
    "            ]\n",
    "            processed_list = current_list[len(previous_ents):]\n",
    "            processed_preds[doc] = processed_list\n",
    "    return processed_preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def print_ent_comparison(evaled_ids, processed_preds, target_entities):\n",
    "    for doc in evaled_ids:\n",
    "        ents = processed_preds[doc]\n",
    "        gold_ents = [e[0] for e in target_entities[doc]]\n",
    "        print('\\n\\n\\nFor doc ', doc, ':')\n",
    "        print('-----------------------------------------')\n",
    "        print('PREDICTIONS:')\n",
    "        print('\\n'.join(ents))\n",
    "        print('-------------------')\n",
    "        print('GOLD STANDARD:')\n",
    "        print('\\n'.join(gold_ents))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "For doc  PMID10707361_abstract :\n",
      "-----------------------------------------\n",
      "PREDICTIONS:\n",
      "MAPK\n",
      "SIPK\n",
      "WIPK\n",
      "NahG\n",
      "salicylic acid-induced protein kinase\n",
      "wounding-induced protein kinase\n",
      "nitric oxide\n",
      "salicylic acid\n",
      "ethylene\n",
      "jasmonic acid\n",
      "tobacco\n",
      "-------------------\n",
      "GOLD STANDARD:\n",
      "nitric oxide\n",
      "salicylic acid (SA)-induced protein kinase\n",
      "NO\n",
      "tobacco\n",
      "SIPK\n",
      "transgenic NahG tobacco\n",
      "wounding-induced protein kinase\n",
      "kinases\n",
      "WIPK\n",
      "mitogen-activated protein ( MAP ) kinases\n",
      "\n",
      "\n",
      "\n",
      "For doc  PMID10972869_abstract :\n",
      "-----------------------------------------\n",
      "PREDICTIONS:\n",
      "NtPat1\n",
      "NtPat2\n",
      "NtPat3\n",
      "phospholipase A2 (PLA2)\n",
      "patatin\n",
      "12-oxophytodienoic acid\n",
      "tobacco mosaic virus\n",
      "NtPat\n",
      "virus-infected leaves\n",
      "-------------------\n",
      "GOLD STANDARD:\n",
      "NtPat genes\n",
      "oxylipins\n",
      "patatin-like phospholipases\n",
      "PLA2 isoforms\n",
      "tobacco patatin-like cDNAs\n",
      "PLA2\n",
      "NtPat proteins\n",
      "tobacco mosaic virus\n",
      "NtPat3\n",
      "Recombinant NtPat1 and NtPat3 enzymes\n",
      "membrane lipids\n",
      "phospholipase A2\n",
      "NtPat1\n",
      "unsaturated fatty acids\n",
      "NtPat2\n",
      "phosphatidylcholine\n",
      "patatin\n",
      "\n",
      "\n",
      "\n",
      "For doc  PMID10072406_abstract :\n",
      "-----------------------------------------\n",
      "PREDICTIONS:\n",
      "N/A\n",
      "oxylipins\n",
      "colneleic acid\n",
      "colnelenic acid\n",
      "Phytophthora infestans\n",
      "-------------------\n",
      "GOLD STANDARD:\n",
      "oxylipins\n",
      "Bintje\n",
      "oxidized fatty acid derivatives\n",
      "Colnelenic acid\n",
      "divinyl ether fatty acids\n",
      "tobacco\n",
      "divinyl ethers\n",
      "P. infestans\n",
      "Colneleic and colnelenic acids\n",
      "Phytophthora infestans\n",
      "colneleic acid\n",
      "colnelenic acid\n",
      "tobacco mosaic virus\n",
      "potato\n",
      "jasmonic acid family\n",
      "Matilda\n",
      "18-carbon divinyl ether fatty acids\n",
      "late-blight pathogen\n",
      "\n",
      "\n",
      "\n",
      "For doc  PMID10890883_abstract :\n",
      "-----------------------------------------\n",
      "PREDICTIONS:\n",
      "NPR1\n",
      "PR-1\n",
      "NPR1\n",
      "salicylic acid (SA)\n",
      "Arabidopsis thaliana\n",
      "Pseudomonas rhizobacteria\n",
      "Pseudomonas syringae pv. tomato (Pst)\n",
      "-------------------\n",
      "GOLD STANDARD:\n",
      "Arabidopsis genotypes\n",
      "Pst\n",
      "Pseudomonas syringae pv . tomato\n",
      "Arabidopsis thaliana\n",
      "Npr1 transcript\n",
      "NPR1\n",
      "Pseudomonas rhizobacteria\n",
      "SAR marker gene PR-1\n",
      "\n",
      "\n",
      "\n",
      "For doc  PMID10476063_abstract :\n",
      "-----------------------------------------\n",
      "PREDICTIONS:\n",
      "pad3-1\n",
      "Pseudomonas syringae\n",
      "Peronospora parasitica\n",
      "Erysiphe orontii\n",
      "Alternaria brassicicola\n",
      "Botrytis cinerea\n",
      "PAD3\n",
      "camalexin\n",
      "salicylate\n",
      "jasmonate\n",
      "Arabidopsis\n",
      "Pseudomonas syringae\n",
      "Peronospora parasitica\n",
      "Erysiphe orontii\n",
      "Alternaria brassicicola\n",
      "Botrytis cinerea\n",
      "-------------------\n",
      "GOLD STANDARD:\n",
      "indole-type phytoalexin camalexin\n",
      "Erysiphe orontii\n",
      "phytoalexin-deficient Arabidopsis mutant pad3 - 1\n",
      "PAD3\n",
      "Botrytis cinerea\n",
      "pad3 - 1\n",
      "Alternaria-challenged pad3 - 1 plants\n",
      "A. brassicicola\n",
      "camalexin\n",
      "salicylate-dependent and jasmonate/ethylene-dependent defense genes\n",
      "Pseudomonas syringae\n",
      "Peronospora parasitica\n",
      "Alternaria brassicicola\n"
     ]
    }
   ],
   "source": [
    "processed_preds_1 = process_preds(evaled_ids, ent_df_1)\n",
    "print_ent_comparison(evaled_ids, processed_preds_1, target_entities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "For doc  PMID10707361_abstract :\n",
      "-----------------------------------------\n",
      "PREDICTIONS:\n",
      "MAPK\n",
      "SIPK\n",
      "WIPK\n",
      "NahG\n",
      "salicylic acid-induced protein kinase\n",
      "wounding-induced protein kinase\n",
      "nitric oxide\n",
      "salicylic acid\n",
      "ethylene\n",
      "jasmonic acid\n",
      "tobacco\n",
      "-------------------\n",
      "GOLD STANDARD:\n",
      "nitric oxide\n",
      "salicylic acid (SA)-induced protein kinase\n",
      "NO\n",
      "tobacco\n",
      "SIPK\n",
      "transgenic NahG tobacco\n",
      "wounding-induced protein kinase\n",
      "kinases\n",
      "WIPK\n",
      "mitogen-activated protein ( MAP ) kinases\n",
      "\n",
      "\n",
      "\n",
      "For doc  PMID10972869_abstract :\n",
      "-----------------------------------------\n",
      "PREDICTIONS:\n",
      "NtPat1\n",
      "NtPat2\n",
      "NtPat3\n",
      "phospholipase A2 (PLA2)\n",
      "patatin\n",
      "12-oxophytodienoic acid\n",
      "unsaturated fatty acids\n",
      "phosphatidylcholine\n",
      "potato\n",
      "NtPat\n",
      "patatin-like proteins\n",
      "plant\n",
      "-------------------\n",
      "GOLD STANDARD:\n",
      "NtPat genes\n",
      "oxylipins\n",
      "patatin-like phospholipases\n",
      "PLA2 isoforms\n",
      "tobacco patatin-like cDNAs\n",
      "PLA2\n",
      "NtPat proteins\n",
      "tobacco mosaic virus\n",
      "NtPat3\n",
      "Recombinant NtPat1 and NtPat3 enzymes\n",
      "membrane lipids\n",
      "phospholipase A2\n",
      "NtPat1\n",
      "unsaturated fatty acids\n",
      "NtPat2\n",
      "phosphatidylcholine\n",
      "patatin\n",
      "\n",
      "\n",
      "\n",
      "For doc  PMID10072406_abstract :\n",
      "-----------------------------------------\n",
      "PREDICTIONS:\n",
      "oxylipins\n",
      "colneleic acid\n",
      "colnelenic acid\n",
      "Phytophthora infestans\n",
      "tobacco mosaic virus\n",
      "-------------------\n",
      "GOLD STANDARD:\n",
      "oxylipins\n",
      "Bintje\n",
      "oxidized fatty acid derivatives\n",
      "Colnelenic acid\n",
      "divinyl ether fatty acids\n",
      "tobacco\n",
      "divinyl ethers\n",
      "P. infestans\n",
      "Colneleic and colnelenic acids\n",
      "Phytophthora infestans\n",
      "colneleic acid\n",
      "colnelenic acid\n",
      "tobacco mosaic virus\n",
      "potato\n",
      "jasmonic acid family\n",
      "Matilda\n",
      "18-carbon divinyl ether fatty acids\n",
      "late-blight pathogen\n",
      "\n",
      "\n",
      "\n",
      "For doc  PMID10890883_abstract :\n",
      "-----------------------------------------\n",
      "PREDICTIONS:\n",
      "NPR1\n",
      "PR-1\n",
      "NPR1\n",
      "Arabidopsis thaliana\n",
      "Pseudomonas syringae pv. tomato\n",
      "SAR\n",
      "ISR\n",
      "-------------------\n",
      "GOLD STANDARD:\n",
      "Arabidopsis genotypes\n",
      "Pst\n",
      "Pseudomonas syringae pv . tomato\n",
      "Arabidopsis thaliana\n",
      "Npr1 transcript\n",
      "NPR1\n",
      "Pseudomonas rhizobacteria\n",
      "SAR marker gene PR-1\n",
      "\n",
      "\n",
      "\n",
      "For doc  PMID10476063_abstract :\n",
      "-----------------------------------------\n",
      "PREDICTIONS:\n",
      "pad3-1\n",
      "PAD3\n",
      "camalexin\n",
      "salicylate\n",
      "jasmonate\n",
      "Arabidopsis\n",
      "Pseudomonas syringae\n",
      "Peronospora parasitica\n",
      "Erysiphe orontii\n",
      "Alternaria brassicicola\n",
      "Botrytis cinerea\n",
      "not found\n",
      "-------------------\n",
      "GOLD STANDARD:\n",
      "indole-type phytoalexin camalexin\n",
      "Erysiphe orontii\n",
      "phytoalexin-deficient Arabidopsis mutant pad3 - 1\n",
      "PAD3\n",
      "Botrytis cinerea\n",
      "pad3 - 1\n",
      "Alternaria-challenged pad3 - 1 plants\n",
      "A. brassicicola\n",
      "camalexin\n",
      "salicylate-dependent and jasmonate/ethylene-dependent defense genes\n",
      "Pseudomonas syringae\n",
      "Peronospora parasitica\n",
      "Alternaria brassicicola\n"
     ]
    }
   ],
   "source": [
    "processed_preds_2 = process_preds(evaled_ids, ent_df_2)\n",
    "print_ent_comparison(evaled_ids, processed_preds_2, target_entities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "For doc  PMID10707361_abstract :\n",
      "-----------------------------------------\n",
      "PREDICTIONS:\n",
      "MAPK\n",
      "SIPK\n",
      "WIPK\n",
      "NahG\n",
      "salicylic acid-induced protein kinase\n",
      "wounding-induced protein kinase\n",
      "nitric oxide\n",
      "salicylic acid\n",
      "ethylene\n",
      "jasmonic acid\n",
      "tobacco\n",
      "-------------------\n",
      "GOLD STANDARD:\n",
      "nitric oxide\n",
      "salicylic acid (SA)-induced protein kinase\n",
      "NO\n",
      "tobacco\n",
      "SIPK\n",
      "transgenic NahG tobacco\n",
      "wounding-induced protein kinase\n",
      "kinases\n",
      "WIPK\n",
      "mitogen-activated protein ( MAP ) kinases\n",
      "\n",
      "\n",
      "\n",
      "For doc  PMID10972869_abstract :\n",
      "-----------------------------------------\n",
      "PREDICTIONS:\n",
      "NtPat1\n",
      "NtPat2\n",
      "NtPat3\n",
      "phospholipase A2 (PLA2)\n",
      "patatin\n",
      "12-oxophytodienoic acid\n",
      "unsaturated fatty acids\n",
      "phosphatidylcholine\n",
      "potato\n",
      "NtPat\n",
      "patatin-like proteins\n",
      "plant\n",
      "-------------------\n",
      "GOLD STANDARD:\n",
      "NtPat genes\n",
      "oxylipins\n",
      "patatin-like phospholipases\n",
      "PLA2 isoforms\n",
      "tobacco patatin-like cDNAs\n",
      "PLA2\n",
      "NtPat proteins\n",
      "tobacco mosaic virus\n",
      "NtPat3\n",
      "Recombinant NtPat1 and NtPat3 enzymes\n",
      "membrane lipids\n",
      "phospholipase A2\n",
      "NtPat1\n",
      "unsaturated fatty acids\n",
      "NtPat2\n",
      "phosphatidylcholine\n",
      "patatin\n",
      "\n",
      "\n",
      "\n",
      "For doc  PMID10072406_abstract :\n",
      "-----------------------------------------\n",
      "PREDICTIONS:\n",
      "oxylipins\n",
      "colneleic acid\n",
      "colnelenic acid\n",
      "Phytophthora infestans\n",
      "tobacco mosaic virus\n",
      "-------------------\n",
      "GOLD STANDARD:\n",
      "oxylipins\n",
      "Bintje\n",
      "oxidized fatty acid derivatives\n",
      "Colnelenic acid\n",
      "divinyl ether fatty acids\n",
      "tobacco\n",
      "divinyl ethers\n",
      "P. infestans\n",
      "Colneleic and colnelenic acids\n",
      "Phytophthora infestans\n",
      "colneleic acid\n",
      "colnelenic acid\n",
      "tobacco mosaic virus\n",
      "potato\n",
      "jasmonic acid family\n",
      "Matilda\n",
      "18-carbon divinyl ether fatty acids\n",
      "late-blight pathogen\n",
      "\n",
      "\n",
      "\n",
      "For doc  PMID10890883_abstract :\n",
      "-----------------------------------------\n",
      "PREDICTIONS:\n",
      "NPR1\n",
      "PR-1\n",
      "NPR1\n",
      "Arabidopsis thaliana\n",
      "Pseudomonas syringae pv. tomato\n",
      "SAR\n",
      "ISR\n",
      "-------------------\n",
      "GOLD STANDARD:\n",
      "Arabidopsis genotypes\n",
      "Pst\n",
      "Pseudomonas syringae pv . tomato\n",
      "Arabidopsis thaliana\n",
      "Npr1 transcript\n",
      "NPR1\n",
      "Pseudomonas rhizobacteria\n",
      "SAR marker gene PR-1\n",
      "\n",
      "\n",
      "\n",
      "For doc  PMID10476063_abstract :\n",
      "-----------------------------------------\n",
      "PREDICTIONS:\n",
      "pad3-1\n",
      "PAD3\n",
      "camalexin\n",
      "salicylate\n",
      "jasmonate\n",
      "Arabidopsis\n",
      "Pseudomonas syringae\n",
      "Peronospora parasitica\n",
      "Erysiphe orontii\n",
      "Alternaria brassicicola\n",
      "Botrytis cinerea\n",
      "not found\n",
      "-------------------\n",
      "GOLD STANDARD:\n",
      "indole-type phytoalexin camalexin\n",
      "Erysiphe orontii\n",
      "phytoalexin-deficient Arabidopsis mutant pad3 - 1\n",
      "PAD3\n",
      "Botrytis cinerea\n",
      "pad3 - 1\n",
      "Alternaria-challenged pad3 - 1 plants\n",
      "A. brassicicola\n",
      "camalexin\n",
      "salicylate-dependent and jasmonate/ethylene-dependent defense genes\n",
      "Pseudomonas syringae\n",
      "Peronospora parasitica\n",
      "Alternaria brassicicola\n"
     ]
    }
   ],
   "source": [
    "processed_preds_3 = process_preds(evaled_ids, ent_df_3)\n",
    "print_ent_comparison(evaled_ids, processed_preds_3, target_entities)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Direct comparison of the entities pulled from each of the three versions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "On document PMID10707361_abstract\n",
      "-----------------------------------------------\n",
      "Iteration 1entities:\n",
      "['MAPK', 'SIPK', 'WIPK', 'NahG', 'salicylic acid-induced protein kinase', 'wounding-induced protein kinase', 'nitric oxide', 'salicylic acid', 'ethylene', 'jasmonic acid', 'tobacco']\n",
      "Iteration 2entities:\n",
      "['MAPK', 'SIPK', 'WIPK', 'NahG', 'salicylic acid-induced protein kinase', 'wounding-induced protein kinase', 'nitric oxide', 'salicylic acid', 'ethylene', 'jasmonic acid', 'tobacco']\n",
      "Iteration 3entities:\n",
      "['MAPK', 'SIPK', 'WIPK', 'NahG', 'salicylic acid-induced protein kinase', 'wounding-induced protein kinase', 'nitric oxide', 'salicylic acid', 'ethylene', 'jasmonic acid', 'tobacco']\n",
      "\n",
      "On document PMID10972869_abstract\n",
      "-----------------------------------------------\n",
      "Iteration 1entities:\n",
      "['NtPat1', 'NtPat2', 'NtPat3', 'phospholipase A2 (PLA2)', 'patatin', '12-oxophytodienoic acid', 'tobacco mosaic virus', 'NtPat', 'virus-infected leaves']\n",
      "Iteration 2entities:\n",
      "['NtPat1', 'NtPat2', 'NtPat3', 'phospholipase A2 (PLA2)', 'patatin', '12-oxophytodienoic acid', 'unsaturated fatty acids', 'phosphatidylcholine', 'potato', 'NtPat', 'patatin-like proteins', 'plant']\n",
      "Iteration 3entities:\n",
      "['NtPat1', 'NtPat2', 'NtPat3', 'phospholipase A2 (PLA2)', 'patatin', '12-oxophytodienoic acid', 'unsaturated fatty acids', 'phosphatidylcholine', 'potato', 'NtPat', 'patatin-like proteins', 'plant']\n",
      "\n",
      "On document PMID10072406_abstract\n",
      "-----------------------------------------------\n",
      "Iteration 1entities:\n",
      "['N/A', 'oxylipins', 'colneleic acid', 'colnelenic acid', 'Phytophthora infestans']\n",
      "Iteration 2entities:\n",
      "['oxylipins', 'colneleic acid', 'colnelenic acid', 'Phytophthora infestans', 'tobacco mosaic virus']\n",
      "Iteration 3entities:\n",
      "['oxylipins', 'colneleic acid', 'colnelenic acid', 'Phytophthora infestans', 'tobacco mosaic virus']\n",
      "\n",
      "On document PMID10890883_abstract\n",
      "-----------------------------------------------\n",
      "Iteration 1entities:\n",
      "['NPR1', 'PR-1', 'NPR1', 'salicylic acid (SA)', 'Arabidopsis thaliana', 'Pseudomonas rhizobacteria', 'Pseudomonas syringae pv. tomato (Pst)']\n",
      "Iteration 2entities:\n",
      "['NPR1', 'PR-1', 'NPR1', 'Arabidopsis thaliana', 'Pseudomonas syringae pv. tomato', 'SAR', 'ISR']\n",
      "Iteration 3entities:\n",
      "['NPR1', 'PR-1', 'NPR1', 'Arabidopsis thaliana', 'Pseudomonas syringae pv. tomato', 'SAR', 'ISR']\n",
      "\n",
      "On document PMID10476063_abstract\n",
      "-----------------------------------------------\n",
      "Iteration 1entities:\n",
      "['pad3-1', 'Pseudomonas syringae', 'Peronospora parasitica', 'Erysiphe orontii', 'Alternaria brassicicola', 'Botrytis cinerea', 'PAD3', 'camalexin', 'salicylate', 'jasmonate', 'Arabidopsis', 'Pseudomonas syringae', 'Peronospora parasitica', 'Erysiphe orontii', 'Alternaria brassicicola', 'Botrytis cinerea']\n",
      "Iteration 2entities:\n",
      "['pad3-1', 'PAD3', 'camalexin', 'salicylate', 'jasmonate', 'Arabidopsis', 'Pseudomonas syringae', 'Peronospora parasitica', 'Erysiphe orontii', 'Alternaria brassicicola', 'Botrytis cinerea', 'not found']\n",
      "Iteration 3entities:\n",
      "['pad3-1', 'PAD3', 'camalexin', 'salicylate', 'jasmonate', 'Arabidopsis', 'Pseudomonas syringae', 'Peronospora parasitica', 'Erysiphe orontii', 'Alternaria brassicicola', 'Botrytis cinerea', 'not found']\n"
     ]
    }
   ],
   "source": [
    "for doc in evaled_ids:\n",
    "    print('\\nOn document', doc)\n",
    "    print('-----------------------------------------------')\n",
    "    for n, preds in {'1': processed_preds_1, '2': processed_preds_2, '3': processed_preds_3}.items():\n",
    "        print('Iteration ' + n + 'entities:')\n",
    "        print(preds[doc])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluating relations\n",
    "The relation types we extracted with OntoGPT don't map well onto the PICKLE relations, so we'll just print the text of the abstract here along with the relations from both PICKLE and the predictions to get an idea of how it's doing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "ent_to_name_1 = ent_df_1[['id', 'name']].set_index('id').to_dict()['name']\n",
    "ent_to_name_2 = ent_df_2[['id', 'name']].set_index('id').to_dict()['name']\n",
    "ent_to_name_3 = ent_df_3[['id', 'name']].set_index('id').to_dict()['name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_rel_comparisons(gold, ent_to_name, rel_df):\n",
    "    for doc in gold:\n",
    "        toks = [tok for sent in doc['sentences'] for tok in sent]\n",
    "        text = ' '.join(toks)\n",
    "        gold_rels = []\n",
    "        for sent in doc['relations']:\n",
    "            for rel in sent:\n",
    "                subj = ' '.join(toks[rel[0]:rel[1]+1])\n",
    "                pred = rel[4]\n",
    "                obj = ' '.join(toks[rel[2]: rel[3]+1])\n",
    "                gold_rels.append((subj, pred, obj))\n",
    "        rels = [(ent_to_name[row.subject], row.predicate, ent_to_name[row.object]) for i, row in rel_df[rel_df['doc_key'] == doc['doc_key']].iterrows()]\n",
    "        print('\\n\\nAbstract', doc['doc_key'])\n",
    "        print(text)\n",
    "        print('\\nGold relations:')\n",
    "        for rel in gold_rels:\n",
    "            print(rel)\n",
    "        print('\\nPredicted relations:')\n",
    "        for rel in rels:\n",
    "            print(rel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Abstract PMID10707361_abstract\n",
      "In tobacco , two mitogen-activated protein ( MAP ) kinases , designated salicylic acid (SA)-induced protein kinase ( SIPK ) and wounding-induced protein kinase ( WIPK ) are activated in a disease resistance-specific manner following pathogen infection or elicitor treatment . To investigate whether nitric oxide ( NO ) , SA , ethylene , or jasmonic acid ( JA ) are involved in this phenomenon , the ability of these defense signals to activate these kinases was assessed . Both NO and SA activated SIPK ; however , they did not activate WIPK . Additional analyses with transgenic NahG tobacco revealed that SA is required for the NO-mediated induction of SIPK . Neither JA nor ethylene activated SIPK or WIPK . Thus , SIPK may function downstream of SA in the NO signaling pathway for defense responses , while the signals responsible for resistance-associated WIPK activation have yet to be determined .\n",
      "\n",
      "Gold relations:\n",
      "('salicylic acid (SA)-induced protein kinase', 'is-in', 'tobacco')\n",
      "('SIPK', 'is-in', 'tobacco')\n",
      "('wounding-induced protein kinase', 'is-in', 'tobacco')\n",
      "('WIPK', 'is-in', 'tobacco')\n",
      "('NO', 'activates', 'SIPK')\n",
      "('SA', 'activates', 'SIPK')\n",
      "('SA', 'activates', 'SIPK')\n",
      "('NO-mediated induction', 'activates', 'SIPK')\n",
      "('SIPK', 'is-in', 'NO signaling pathway')\n",
      "\n",
      "Predicted relations:\n",
      "\n",
      "\n",
      "Abstract PMID10972869_abstract\n",
      "Recent evidence suggests that oxidized lipid-derived molecules play significant roles in inducible plant defence responses against microbial pathogens , either by directly deterring parasite multiplication , or as signals involved in the induction of sets of defence genes . The synthesis of these oxylipins was hypothesized to be initiated by the phospholipase A2-mediated release of unsaturated fatty acids from membrane lipids . Here , we demonstrate that , in tobacco leaves reacting hypersensitively to tobacco mosaic virus , a strong increase in soluble phospholipase A2 ( PLA2 ) activity occurs at the onset of necrotic lesion appearance . This rapid PLA2 activation occurred before the accumulation of 12-oxophytodienoic and jasmonic acids , two fatty acid-derived defence signals . Three PLA2 isoforms were separated and the most active enzyme was partially purified , its N-terminal sequence displaying similarity with patatin , the major storage protein in potato tubers . Three related tobacco patatin-like cDNAs , called NtPat1 , NtPat2 and NtPat3 , were cloned , with NtPat2 encoding the PLA2 isolated from infected leaves . RT-PCR experiments showed a rapid transcriptional activation of the three NtPat genes in virus-infected leaves , preceding the increase in PLA2 activity . Recombinant NtPat1 and NtPat3 enzymes were active in an assay using labelled bacterial membranes , and also displayed high bona fide PLA2 activity on phosphatidylcholine substrate . These results point to a possible new role of patatin-like phospholipases in inducible plant defence responses . The induction kinetics together with the enzymatic activity data indicate that the NtPat proteins may provide precursors for oxylipin synthesis during the hypersensitive response to pathogens .\n",
      "\n",
      "Gold relations:\n",
      "('unsaturated fatty acids', 'is-in', 'membrane lipids')\n",
      "('tobacco mosaic virus', 'inhibits', 'tobacco leaves')\n",
      "('tobacco mosaic virus', 'activates', 'phospholipase A2')\n",
      "('tobacco mosaic virus', 'activates', 'PLA2')\n",
      "('patatin', 'is-in', 'potato tubers')\n",
      "('NtPat2', 'produces', 'PLA2')\n",
      "('NtPat proteins', 'interacts', 'oxylipin synthesis')\n",
      "\n",
      "Predicted relations:\n",
      "('NtPat2', 'GeneProteinInteraction', 'phospholipase A2 (PLA2)')\n",
      "('NtPat', 'GeneOrganismRelationship', 'virus-infected leaves')\n",
      "\n",
      "\n",
      "Abstract PMID10890883_abstract\n",
      "The plant-signaling molecules salicylic acid ( SA ) and jasmonic acid ( JA ) play an important role in induced disease resistance pathways . Cross-talk between SA- and JA-dependent pathways can result in inhibition of JA-mediated defense responses . We investigated possible antagonistic interactions between the SA-dependent systemic acquired resistance ( SAR ) pathway , which is induced upon pathogen infection , and the JA-dependent induced systemic resistance ( ISR ) pathway , which is triggered by nonpathogenic Pseudomonas rhizobacteria . In Arabidopsis thaliana , SAR and ISR are effective against a broad spectrum of pathogens , including the foliar pathogen Pseudomonas syringae pv . tomato ( Pst ) . Simultaneous activation of SAR and ISR resulted in an additive effect on the level of induced protection against Pst . In Arabidopsis genotypes that are blocked in either SAR or ISR , this additive effect was not evident . Moreover , induction of ISR did not affect the expression of the SAR marker gene PR-1 in plants expressing SAR . Together , these observations demonstrate that the SAR and the ISR pathway are compatible and that there is no significant cross-talk between these pathways . SAR and ISR both require the key regulatory protein NPR1 . Plants expressing both types of induced resistance did not show elevated Npr1 transcript levels , indicating that the constitutive level of NPR1 is sufficient to facilitate simultaneous expression of SAR and ISR . These results suggest that the enhanced level of protection is established through parallel activation of complementary , NPR1-dependent defense responses that are both active against Pst . Therefore , combining SAR and ISR provides an attractive tool for the improvement of disease control .\n",
      "\n",
      "Gold relations:\n",
      "('SA- and JA-dependent pathways', 'inhibits', 'JA-mediated defense responses')\n",
      "('Pseudomonas rhizobacteria', 'activates', 'JA-dependent induced systemic resistance ( ISR ) pathway')\n",
      "('SAR', 'is-in', 'Arabidopsis thaliana')\n",
      "('ISR', 'is-in', 'Arabidopsis thaliana')\n",
      "('ISR', 'inhibits', 'Pst')\n",
      "('ISR', 'inhibits', 'Pseudomonas syringae pv . tomato')\n",
      "('SAR', 'inhibits', 'Pseudomonas syringae pv . tomato')\n",
      "('SAR', 'inhibits', 'Pst')\n",
      "('SAR', 'inhibits', 'Pst')\n",
      "('ISR', 'inhibits', 'Pst')\n",
      "('NPR1', 'interacts', 'ISR')\n",
      "('NPR1', 'interacts', 'SAR')\n",
      "('NPR1', 'activates', 'SAR')\n",
      "('NPR1', 'activates', 'ISR')\n",
      "('NPR1-dependent defense responses', 'inhibits', 'Pst')\n",
      "\n",
      "Predicted relations:\n",
      "('NPR1', 'GeneProteinInteraction', 'PR-1')\n",
      "\n",
      "\n",
      "Abstract PMID10072406_abstract\n",
      "We conducted a study of the patterns and dynamics of oxidized fatty acid derivatives ( oxylipins ) in potato leaves infected with the late-blight pathogen Phytophthora infestans . Two 18-carbon divinyl ether fatty acids , colneleic acid and colnelenic acid , accumulated during disease development . To date , there are no reports that such compounds have been detected in higher plants . The divinyl ether fatty acids accumulate more rapidly in potato cultivar Matilda ( a cultivar with increased resistance to late blight ) than in cultivar Bintje , a susceptible cultivar . Colnelenic acid reached levels of up to approximately 24 nmol ( 7 microgram ) per g fresh weight of tissue in infected leaves . By contrast , levels of members of the jasmonic acid family did not change significantly during pathogenesis . The divinyl ethers also accumulated during the incompatible interaction of tobacco with tobacco mosaic virus . Colneleic and colnelenic acids were found to be inhibitory to P. infestans , suggesting a function in plant defense for divinyl ethers , which are unstable compounds rarely encountered in biological systems .\n",
      "\n",
      "Gold relations:\n",
      "('oxidized fatty acid derivatives', 'is-in', 'potato leaves')\n",
      "('oxylipins', 'is-in', 'potato leaves')\n",
      "('potato leaves', 'interacts', 'late-blight pathogen')\n",
      "('potato leaves', 'interacts', 'Phytophthora infestans')\n",
      "('Matilda', 'inhibits', 'late blight')\n",
      "('divinyl ether fatty acids', 'is-in', 'Matilda')\n",
      "('divinyl ether fatty acids', 'is-in', 'Bintje')\n",
      "('tobacco', 'interacts', 'tobacco mosaic virus')\n",
      "('tobacco mosaic virus', 'activates', 'divinyl ethers')\n",
      "('Colneleic and colnelenic acids', 'inhibits', 'P. infestans')\n",
      "\n",
      "Predicted relations:\n",
      "('N/A', 'GeneGeneInteraction', 'N/A')\n",
      "('N/A', 'GeneProteinInteraction', 'N/A')\n",
      "('N/A', 'GeneOrganismRelationship', 'N/A')\n",
      "('N/A', 'ProteinProteinInteraction', 'N/A')\n",
      "('N/A', 'ProteinOrganismRelationship', 'N/A')\n",
      "('N/A', 'GeneMoleculeInteraction', 'N/A')\n",
      "('N/A', 'ProteinMoleculeInteraction', 'N/A')\n",
      "\n",
      "\n",
      "Abstract PMID10476063_abstract\n",
      "The phytoalexin-deficient Arabidopsis mutant pad3 - 1 , which is affected in the production of the indole-type phytoalexin camalexin , has previously been shown not to display altered susceptibility to either the bacterium Pseudomonas syringae ( Glazebrook & Ausubel 1994 ; Proc . Natl . Acad . Sci . USA , 91 : 8955 - 8959 ) or the biotrophic fungi Peronospora parasitica ( Glazebrook et al. 1997 ; Genetics , 146 : 381 - 392 ) and Erysiphe orontii ( Reuber et al. 1998 ; Plant J. 16 : 473 - 485 ) . We now show that this mutant is markedly more susceptible than its wild-type parental line to infection by the necrotrophic fungus Alternaria brassicicola , but not to Botrytis cinerea . A strong camalexin response was elicited in wild-type plants inoculated with either Alternaria brassicicola or Botrytis cinerea , whereas no camalexin could be detected in pad3 - 1 challenged with these fungi . Hence , PAD3 appears to be a key determinant in resistance to at least A. brassicicola . The induction of salicylate-dependent and jasmonate/ethylene-dependent defense genes was not reduced in Alternaria-challenged pad3 - 1 plants compared to similarly treated wild-type plants . Camalexin production could not be triggered by exogenous application of either salicylate , ethylene or jasmonate and was not , or not strongly , reduced in mutants with defects in perception of these defense-related signal molecules . Camalexin-production appears to be controlled by a pathway that exhibits little cross-talk with salicylate- , ethylene- and jasmonate-dependent signalling events .\n",
      "\n",
      "Gold relations:\n",
      "('Alternaria brassicicola', 'activates', 'camalexin response')\n",
      "('Botrytis cinerea', 'activates', 'camalexin response')\n",
      "('PAD3', 'inhibits', 'A. brassicicola')\n",
      "('salicylate-dependent and jasmonate/ethylene-dependent defense genes', 'is-in', 'Alternaria-challenged pad3 - 1 plants')\n",
      "\n",
      "Predicted relations:\n"
     ]
    }
   ],
   "source": [
    "print_rel_comparisons(gold, ent_to_name_1, rel_df_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Abstract PMID10707361_abstract\n",
      "In tobacco , two mitogen-activated protein ( MAP ) kinases , designated salicylic acid (SA)-induced protein kinase ( SIPK ) and wounding-induced protein kinase ( WIPK ) are activated in a disease resistance-specific manner following pathogen infection or elicitor treatment . To investigate whether nitric oxide ( NO ) , SA , ethylene , or jasmonic acid ( JA ) are involved in this phenomenon , the ability of these defense signals to activate these kinases was assessed . Both NO and SA activated SIPK ; however , they did not activate WIPK . Additional analyses with transgenic NahG tobacco revealed that SA is required for the NO-mediated induction of SIPK . Neither JA nor ethylene activated SIPK or WIPK . Thus , SIPK may function downstream of SA in the NO signaling pathway for defense responses , while the signals responsible for resistance-associated WIPK activation have yet to be determined .\n",
      "\n",
      "Gold relations:\n",
      "('salicylic acid (SA)-induced protein kinase', 'is-in', 'tobacco')\n",
      "('SIPK', 'is-in', 'tobacco')\n",
      "('wounding-induced protein kinase', 'is-in', 'tobacco')\n",
      "('WIPK', 'is-in', 'tobacco')\n",
      "('NO', 'activates', 'SIPK')\n",
      "('SA', 'activates', 'SIPK')\n",
      "('SA', 'activates', 'SIPK')\n",
      "('NO-mediated induction', 'activates', 'SIPK')\n",
      "('SIPK', 'is-in', 'NO signaling pathway')\n",
      "\n",
      "Predicted relations:\n",
      "('SIPK', 'GeneMoleculeInteraction', 'nitric oxide')\n",
      "('SIPK', 'GeneMoleculeInteraction', 'salicylic acid')\n",
      "\n",
      "\n",
      "Abstract PMID10972869_abstract\n",
      "Recent evidence suggests that oxidized lipid-derived molecules play significant roles in inducible plant defence responses against microbial pathogens , either by directly deterring parasite multiplication , or as signals involved in the induction of sets of defence genes . The synthesis of these oxylipins was hypothesized to be initiated by the phospholipase A2-mediated release of unsaturated fatty acids from membrane lipids . Here , we demonstrate that , in tobacco leaves reacting hypersensitively to tobacco mosaic virus , a strong increase in soluble phospholipase A2 ( PLA2 ) activity occurs at the onset of necrotic lesion appearance . This rapid PLA2 activation occurred before the accumulation of 12-oxophytodienoic and jasmonic acids , two fatty acid-derived defence signals . Three PLA2 isoforms were separated and the most active enzyme was partially purified , its N-terminal sequence displaying similarity with patatin , the major storage protein in potato tubers . Three related tobacco patatin-like cDNAs , called NtPat1 , NtPat2 and NtPat3 , were cloned , with NtPat2 encoding the PLA2 isolated from infected leaves . RT-PCR experiments showed a rapid transcriptional activation of the three NtPat genes in virus-infected leaves , preceding the increase in PLA2 activity . Recombinant NtPat1 and NtPat3 enzymes were active in an assay using labelled bacterial membranes , and also displayed high bona fide PLA2 activity on phosphatidylcholine substrate . These results point to a possible new role of patatin-like phospholipases in inducible plant defence responses . The induction kinetics together with the enzymatic activity data indicate that the NtPat proteins may provide precursors for oxylipin synthesis during the hypersensitive response to pathogens .\n",
      "\n",
      "Gold relations:\n",
      "('unsaturated fatty acids', 'is-in', 'membrane lipids')\n",
      "('tobacco mosaic virus', 'inhibits', 'tobacco leaves')\n",
      "('tobacco mosaic virus', 'activates', 'phospholipase A2')\n",
      "('tobacco mosaic virus', 'activates', 'PLA2')\n",
      "('patatin', 'is-in', 'potato tubers')\n",
      "('NtPat2', 'produces', 'PLA2')\n",
      "('NtPat proteins', 'interacts', 'oxylipin synthesis')\n",
      "\n",
      "Predicted relations:\n",
      "('NtPat2', 'GeneProteinInteraction', 'phospholipase A2 (PLA2)')\n",
      "('NtPat', 'GeneOrganismRelationship', 'tobacco')\n",
      "('patatin-like proteins', 'ProteinOrganismRelationship', 'plant')\n",
      "\n",
      "\n",
      "Abstract PMID10890883_abstract\n",
      "The plant-signaling molecules salicylic acid ( SA ) and jasmonic acid ( JA ) play an important role in induced disease resistance pathways . Cross-talk between SA- and JA-dependent pathways can result in inhibition of JA-mediated defense responses . We investigated possible antagonistic interactions between the SA-dependent systemic acquired resistance ( SAR ) pathway , which is induced upon pathogen infection , and the JA-dependent induced systemic resistance ( ISR ) pathway , which is triggered by nonpathogenic Pseudomonas rhizobacteria . In Arabidopsis thaliana , SAR and ISR are effective against a broad spectrum of pathogens , including the foliar pathogen Pseudomonas syringae pv . tomato ( Pst ) . Simultaneous activation of SAR and ISR resulted in an additive effect on the level of induced protection against Pst . In Arabidopsis genotypes that are blocked in either SAR or ISR , this additive effect was not evident . Moreover , induction of ISR did not affect the expression of the SAR marker gene PR-1 in plants expressing SAR . Together , these observations demonstrate that the SAR and the ISR pathway are compatible and that there is no significant cross-talk between these pathways . SAR and ISR both require the key regulatory protein NPR1 . Plants expressing both types of induced resistance did not show elevated Npr1 transcript levels , indicating that the constitutive level of NPR1 is sufficient to facilitate simultaneous expression of SAR and ISR . These results suggest that the enhanced level of protection is established through parallel activation of complementary , NPR1-dependent defense responses that are both active against Pst . Therefore , combining SAR and ISR provides an attractive tool for the improvement of disease control .\n",
      "\n",
      "Gold relations:\n",
      "('SA- and JA-dependent pathways', 'inhibits', 'JA-mediated defense responses')\n",
      "('Pseudomonas rhizobacteria', 'activates', 'JA-dependent induced systemic resistance ( ISR ) pathway')\n",
      "('SAR', 'is-in', 'Arabidopsis thaliana')\n",
      "('ISR', 'is-in', 'Arabidopsis thaliana')\n",
      "('ISR', 'inhibits', 'Pst')\n",
      "('ISR', 'inhibits', 'Pseudomonas syringae pv . tomato')\n",
      "('SAR', 'inhibits', 'Pseudomonas syringae pv . tomato')\n",
      "('SAR', 'inhibits', 'Pst')\n",
      "('SAR', 'inhibits', 'Pst')\n",
      "('ISR', 'inhibits', 'Pst')\n",
      "('NPR1', 'interacts', 'ISR')\n",
      "('NPR1', 'interacts', 'SAR')\n",
      "('NPR1', 'activates', 'SAR')\n",
      "('NPR1', 'activates', 'ISR')\n",
      "('NPR1-dependent defense responses', 'inhibits', 'Pst')\n",
      "\n",
      "Predicted relations:\n",
      "('NPR1', 'GeneProteinInteraction', 'PR-1')\n",
      "('SAR', 'GeneOrganismRelationship', 'Arabidopsis thaliana')\n",
      "('ISR', 'GeneOrganismRelationship', 'Arabidopsis thaliana')\n",
      "('NPR1', 'ProteinProteinInteraction', 'NPR1')\n",
      "('NPR1', 'ProteinOrganismRelationship', 'Arabidopsis thaliana')\n",
      "\n",
      "\n",
      "Abstract PMID10072406_abstract\n",
      "We conducted a study of the patterns and dynamics of oxidized fatty acid derivatives ( oxylipins ) in potato leaves infected with the late-blight pathogen Phytophthora infestans . Two 18-carbon divinyl ether fatty acids , colneleic acid and colnelenic acid , accumulated during disease development . To date , there are no reports that such compounds have been detected in higher plants . The divinyl ether fatty acids accumulate more rapidly in potato cultivar Matilda ( a cultivar with increased resistance to late blight ) than in cultivar Bintje , a susceptible cultivar . Colnelenic acid reached levels of up to approximately 24 nmol ( 7 microgram ) per g fresh weight of tissue in infected leaves . By contrast , levels of members of the jasmonic acid family did not change significantly during pathogenesis . The divinyl ethers also accumulated during the incompatible interaction of tobacco with tobacco mosaic virus . Colneleic and colnelenic acids were found to be inhibitory to P. infestans , suggesting a function in plant defense for divinyl ethers , which are unstable compounds rarely encountered in biological systems .\n",
      "\n",
      "Gold relations:\n",
      "('oxidized fatty acid derivatives', 'is-in', 'potato leaves')\n",
      "('oxylipins', 'is-in', 'potato leaves')\n",
      "('potato leaves', 'interacts', 'late-blight pathogen')\n",
      "('potato leaves', 'interacts', 'Phytophthora infestans')\n",
      "('Matilda', 'inhibits', 'late blight')\n",
      "('divinyl ether fatty acids', 'is-in', 'Matilda')\n",
      "('divinyl ether fatty acids', 'is-in', 'Bintje')\n",
      "('tobacco', 'interacts', 'tobacco mosaic virus')\n",
      "('tobacco mosaic virus', 'activates', 'divinyl ethers')\n",
      "('Colneleic and colnelenic acids', 'inhibits', 'P. infestans')\n",
      "\n",
      "Predicted relations:\n",
      "\n",
      "\n",
      "Abstract PMID10476063_abstract\n",
      "The phytoalexin-deficient Arabidopsis mutant pad3 - 1 , which is affected in the production of the indole-type phytoalexin camalexin , has previously been shown not to display altered susceptibility to either the bacterium Pseudomonas syringae ( Glazebrook & Ausubel 1994 ; Proc . Natl . Acad . Sci . USA , 91 : 8955 - 8959 ) or the biotrophic fungi Peronospora parasitica ( Glazebrook et al. 1997 ; Genetics , 146 : 381 - 392 ) and Erysiphe orontii ( Reuber et al. 1998 ; Plant J. 16 : 473 - 485 ) . We now show that this mutant is markedly more susceptible than its wild-type parental line to infection by the necrotrophic fungus Alternaria brassicicola , but not to Botrytis cinerea . A strong camalexin response was elicited in wild-type plants inoculated with either Alternaria brassicicola or Botrytis cinerea , whereas no camalexin could be detected in pad3 - 1 challenged with these fungi . Hence , PAD3 appears to be a key determinant in resistance to at least A. brassicicola . The induction of salicylate-dependent and jasmonate/ethylene-dependent defense genes was not reduced in Alternaria-challenged pad3 - 1 plants compared to similarly treated wild-type plants . Camalexin production could not be triggered by exogenous application of either salicylate , ethylene or jasmonate and was not , or not strongly , reduced in mutants with defects in perception of these defense-related signal molecules . Camalexin-production appears to be controlled by a pathway that exhibits little cross-talk with salicylate- , ethylene- and jasmonate-dependent signalling events .\n",
      "\n",
      "Gold relations:\n",
      "('Alternaria brassicicola', 'activates', 'camalexin response')\n",
      "('Botrytis cinerea', 'activates', 'camalexin response')\n",
      "('PAD3', 'inhibits', 'A. brassicicola')\n",
      "('salicylate-dependent and jasmonate/ethylene-dependent defense genes', 'is-in', 'Alternaria-challenged pad3 - 1 plants')\n",
      "\n",
      "Predicted relations:\n",
      "('not found', 'GeneGeneInteraction', 'not found')\n",
      "('not found', 'GeneProteinInteraction', 'not found')\n",
      "('not found', 'ProteinProteinInteraction', 'not found')\n",
      "('pad3-1', 'GeneMoleculeInteraction', 'camalexin')\n",
      "('PAD3', 'ProteinMoleculeInteraction', 'camalexin')\n"
     ]
    }
   ],
   "source": [
    "print_rel_comparisons(gold, ent_to_name_2, rel_df_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Abstract PMID10707361_abstract\n",
      "In tobacco , two mitogen-activated protein ( MAP ) kinases , designated salicylic acid (SA)-induced protein kinase ( SIPK ) and wounding-induced protein kinase ( WIPK ) are activated in a disease resistance-specific manner following pathogen infection or elicitor treatment . To investigate whether nitric oxide ( NO ) , SA , ethylene , or jasmonic acid ( JA ) are involved in this phenomenon , the ability of these defense signals to activate these kinases was assessed . Both NO and SA activated SIPK ; however , they did not activate WIPK . Additional analyses with transgenic NahG tobacco revealed that SA is required for the NO-mediated induction of SIPK . Neither JA nor ethylene activated SIPK or WIPK . Thus , SIPK may function downstream of SA in the NO signaling pathway for defense responses , while the signals responsible for resistance-associated WIPK activation have yet to be determined .\n",
      "\n",
      "Gold relations:\n",
      "('salicylic acid (SA)-induced protein kinase', 'is-in', 'tobacco')\n",
      "('SIPK', 'is-in', 'tobacco')\n",
      "('wounding-induced protein kinase', 'is-in', 'tobacco')\n",
      "('WIPK', 'is-in', 'tobacco')\n",
      "('NO', 'activates', 'SIPK')\n",
      "('SA', 'activates', 'SIPK')\n",
      "('SA', 'activates', 'SIPK')\n",
      "('NO-mediated induction', 'activates', 'SIPK')\n",
      "('SIPK', 'is-in', 'NO signaling pathway')\n",
      "\n",
      "Predicted relations:\n",
      "('SIPK', 'GeneMoleculeInteraction', 'nitric oxide')\n",
      "('SIPK', 'GeneMoleculeInteraction', 'salicylic acid')\n",
      "\n",
      "\n",
      "Abstract PMID10972869_abstract\n",
      "Recent evidence suggests that oxidized lipid-derived molecules play significant roles in inducible plant defence responses against microbial pathogens , either by directly deterring parasite multiplication , or as signals involved in the induction of sets of defence genes . The synthesis of these oxylipins was hypothesized to be initiated by the phospholipase A2-mediated release of unsaturated fatty acids from membrane lipids . Here , we demonstrate that , in tobacco leaves reacting hypersensitively to tobacco mosaic virus , a strong increase in soluble phospholipase A2 ( PLA2 ) activity occurs at the onset of necrotic lesion appearance . This rapid PLA2 activation occurred before the accumulation of 12-oxophytodienoic and jasmonic acids , two fatty acid-derived defence signals . Three PLA2 isoforms were separated and the most active enzyme was partially purified , its N-terminal sequence displaying similarity with patatin , the major storage protein in potato tubers . Three related tobacco patatin-like cDNAs , called NtPat1 , NtPat2 and NtPat3 , were cloned , with NtPat2 encoding the PLA2 isolated from infected leaves . RT-PCR experiments showed a rapid transcriptional activation of the three NtPat genes in virus-infected leaves , preceding the increase in PLA2 activity . Recombinant NtPat1 and NtPat3 enzymes were active in an assay using labelled bacterial membranes , and also displayed high bona fide PLA2 activity on phosphatidylcholine substrate . These results point to a possible new role of patatin-like phospholipases in inducible plant defence responses . The induction kinetics together with the enzymatic activity data indicate that the NtPat proteins may provide precursors for oxylipin synthesis during the hypersensitive response to pathogens .\n",
      "\n",
      "Gold relations:\n",
      "('unsaturated fatty acids', 'is-in', 'membrane lipids')\n",
      "('tobacco mosaic virus', 'inhibits', 'tobacco leaves')\n",
      "('tobacco mosaic virus', 'activates', 'phospholipase A2')\n",
      "('tobacco mosaic virus', 'activates', 'PLA2')\n",
      "('patatin', 'is-in', 'potato tubers')\n",
      "('NtPat2', 'produces', 'PLA2')\n",
      "('NtPat proteins', 'interacts', 'oxylipin synthesis')\n",
      "\n",
      "Predicted relations:\n",
      "('NtPat2', 'GeneProteinInteraction', 'phospholipase A2 (PLA2)')\n",
      "('NtPat', 'GeneOrganismRelationship', 'tobacco')\n",
      "('patatin-like proteins', 'ProteinOrganismRelationship', 'plant')\n",
      "\n",
      "\n",
      "Abstract PMID10890883_abstract\n",
      "The plant-signaling molecules salicylic acid ( SA ) and jasmonic acid ( JA ) play an important role in induced disease resistance pathways . Cross-talk between SA- and JA-dependent pathways can result in inhibition of JA-mediated defense responses . We investigated possible antagonistic interactions between the SA-dependent systemic acquired resistance ( SAR ) pathway , which is induced upon pathogen infection , and the JA-dependent induced systemic resistance ( ISR ) pathway , which is triggered by nonpathogenic Pseudomonas rhizobacteria . In Arabidopsis thaliana , SAR and ISR are effective against a broad spectrum of pathogens , including the foliar pathogen Pseudomonas syringae pv . tomato ( Pst ) . Simultaneous activation of SAR and ISR resulted in an additive effect on the level of induced protection against Pst . In Arabidopsis genotypes that are blocked in either SAR or ISR , this additive effect was not evident . Moreover , induction of ISR did not affect the expression of the SAR marker gene PR-1 in plants expressing SAR . Together , these observations demonstrate that the SAR and the ISR pathway are compatible and that there is no significant cross-talk between these pathways . SAR and ISR both require the key regulatory protein NPR1 . Plants expressing both types of induced resistance did not show elevated Npr1 transcript levels , indicating that the constitutive level of NPR1 is sufficient to facilitate simultaneous expression of SAR and ISR . These results suggest that the enhanced level of protection is established through parallel activation of complementary , NPR1-dependent defense responses that are both active against Pst . Therefore , combining SAR and ISR provides an attractive tool for the improvement of disease control .\n",
      "\n",
      "Gold relations:\n",
      "('SA- and JA-dependent pathways', 'inhibits', 'JA-mediated defense responses')\n",
      "('Pseudomonas rhizobacteria', 'activates', 'JA-dependent induced systemic resistance ( ISR ) pathway')\n",
      "('SAR', 'is-in', 'Arabidopsis thaliana')\n",
      "('ISR', 'is-in', 'Arabidopsis thaliana')\n",
      "('ISR', 'inhibits', 'Pst')\n",
      "('ISR', 'inhibits', 'Pseudomonas syringae pv . tomato')\n",
      "('SAR', 'inhibits', 'Pseudomonas syringae pv . tomato')\n",
      "('SAR', 'inhibits', 'Pst')\n",
      "('SAR', 'inhibits', 'Pst')\n",
      "('ISR', 'inhibits', 'Pst')\n",
      "('NPR1', 'interacts', 'ISR')\n",
      "('NPR1', 'interacts', 'SAR')\n",
      "('NPR1', 'activates', 'SAR')\n",
      "('NPR1', 'activates', 'ISR')\n",
      "('NPR1-dependent defense responses', 'inhibits', 'Pst')\n",
      "\n",
      "Predicted relations:\n",
      "('NPR1', 'GeneProteinInteraction', 'PR-1')\n",
      "('SAR', 'GeneOrganismRelationship', 'Arabidopsis thaliana')\n",
      "('ISR', 'GeneOrganismRelationship', 'Arabidopsis thaliana')\n",
      "('NPR1', 'ProteinProteinInteraction', 'NPR1')\n",
      "('NPR1', 'ProteinOrganismRelationship', 'Arabidopsis thaliana')\n",
      "\n",
      "\n",
      "Abstract PMID10072406_abstract\n",
      "We conducted a study of the patterns and dynamics of oxidized fatty acid derivatives ( oxylipins ) in potato leaves infected with the late-blight pathogen Phytophthora infestans . Two 18-carbon divinyl ether fatty acids , colneleic acid and colnelenic acid , accumulated during disease development . To date , there are no reports that such compounds have been detected in higher plants . The divinyl ether fatty acids accumulate more rapidly in potato cultivar Matilda ( a cultivar with increased resistance to late blight ) than in cultivar Bintje , a susceptible cultivar . Colnelenic acid reached levels of up to approximately 24 nmol ( 7 microgram ) per g fresh weight of tissue in infected leaves . By contrast , levels of members of the jasmonic acid family did not change significantly during pathogenesis . The divinyl ethers also accumulated during the incompatible interaction of tobacco with tobacco mosaic virus . Colneleic and colnelenic acids were found to be inhibitory to P. infestans , suggesting a function in plant defense for divinyl ethers , which are unstable compounds rarely encountered in biological systems .\n",
      "\n",
      "Gold relations:\n",
      "('oxidized fatty acid derivatives', 'is-in', 'potato leaves')\n",
      "('oxylipins', 'is-in', 'potato leaves')\n",
      "('potato leaves', 'interacts', 'late-blight pathogen')\n",
      "('potato leaves', 'interacts', 'Phytophthora infestans')\n",
      "('Matilda', 'inhibits', 'late blight')\n",
      "('divinyl ether fatty acids', 'is-in', 'Matilda')\n",
      "('divinyl ether fatty acids', 'is-in', 'Bintje')\n",
      "('tobacco', 'interacts', 'tobacco mosaic virus')\n",
      "('tobacco mosaic virus', 'activates', 'divinyl ethers')\n",
      "('Colneleic and colnelenic acids', 'inhibits', 'P. infestans')\n",
      "\n",
      "Predicted relations:\n",
      "\n",
      "\n",
      "Abstract PMID10476063_abstract\n",
      "The phytoalexin-deficient Arabidopsis mutant pad3 - 1 , which is affected in the production of the indole-type phytoalexin camalexin , has previously been shown not to display altered susceptibility to either the bacterium Pseudomonas syringae ( Glazebrook & Ausubel 1994 ; Proc . Natl . Acad . Sci . USA , 91 : 8955 - 8959 ) or the biotrophic fungi Peronospora parasitica ( Glazebrook et al. 1997 ; Genetics , 146 : 381 - 392 ) and Erysiphe orontii ( Reuber et al. 1998 ; Plant J. 16 : 473 - 485 ) . We now show that this mutant is markedly more susceptible than its wild-type parental line to infection by the necrotrophic fungus Alternaria brassicicola , but not to Botrytis cinerea . A strong camalexin response was elicited in wild-type plants inoculated with either Alternaria brassicicola or Botrytis cinerea , whereas no camalexin could be detected in pad3 - 1 challenged with these fungi . Hence , PAD3 appears to be a key determinant in resistance to at least A. brassicicola . The induction of salicylate-dependent and jasmonate/ethylene-dependent defense genes was not reduced in Alternaria-challenged pad3 - 1 plants compared to similarly treated wild-type plants . Camalexin production could not be triggered by exogenous application of either salicylate , ethylene or jasmonate and was not , or not strongly , reduced in mutants with defects in perception of these defense-related signal molecules . Camalexin-production appears to be controlled by a pathway that exhibits little cross-talk with salicylate- , ethylene- and jasmonate-dependent signalling events .\n",
      "\n",
      "Gold relations:\n",
      "('Alternaria brassicicola', 'activates', 'camalexin response')\n",
      "('Botrytis cinerea', 'activates', 'camalexin response')\n",
      "('PAD3', 'inhibits', 'A. brassicicola')\n",
      "('salicylate-dependent and jasmonate/ethylene-dependent defense genes', 'is-in', 'Alternaria-challenged pad3 - 1 plants')\n",
      "\n",
      "Predicted relations:\n",
      "('not found', 'GeneGeneInteraction', 'not found')\n",
      "('not found', 'GeneOrganismRelationship', 'not found')\n",
      "('not found', 'ProteinOrganismRelationship', 'not found')\n",
      "('pad3-1', 'GeneMoleculeInteraction', 'camalexin')\n",
      "('PAD3', 'ProteinMoleculeInteraction', 'camalexin')\n"
     ]
    }
   ],
   "source": [
    "print_rel_comparisons(gold, ent_to_name_3, rel_df_3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Direct comparison of the relations pulled from each of the three versions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "On document PMID10707361_abstract\n",
      "-----------------------------------------------\n",
      "Iteration 1relations:\n",
      "Iteration 2relations:\n",
      "('AUTO:SIPK', 'GeneMoleculeInteraction', 'CHEBI:16480')\n",
      "('AUTO:SIPK', 'GeneMoleculeInteraction', 'CHEBI:16914')\n",
      "Iteration 3relations:\n",
      "('AUTO:SIPK', 'GeneMoleculeInteraction', 'CHEBI:16480')\n",
      "('AUTO:SIPK', 'GeneMoleculeInteraction', 'CHEBI:16914')\n",
      "\n",
      "On document PMID10972869_abstract\n",
      "-----------------------------------------------\n",
      "Iteration 1relations:\n",
      "('AUTO:NtPat2', 'GeneProteinInteraction', 'PR:000012798')\n",
      "('AUTO:NtPat', 'GeneOrganismRelationship', 'AUTO:virus-infected%20leaves')\n",
      "Iteration 2relations:\n",
      "('AUTO:NtPat2', 'GeneProteinInteraction', 'PR:000012798')\n",
      "('AUTO:NtPat', 'GeneOrganismRelationship', 'NCBITaxon:4097')\n",
      "('AUTO:patatin-like%20proteins', 'ProteinOrganismRelationship', 'AUTO:plant')\n",
      "Iteration 3relations:\n",
      "('AUTO:NtPat2', 'GeneProteinInteraction', 'PR:000012798')\n",
      "('AUTO:NtPat', 'GeneOrganismRelationship', 'NCBITaxon:4097')\n",
      "('AUTO:patatin-like%20proteins', 'ProteinOrganismRelationship', 'AUTO:plant')\n",
      "\n",
      "On document PMID10072406_abstract\n",
      "-----------------------------------------------\n",
      "Iteration 1relations:\n",
      "('AUTO:N/A', 'GeneGeneInteraction', 'AUTO:N/A')\n",
      "('AUTO:N/A', 'GeneProteinInteraction', 'AUTO:N/A')\n",
      "('AUTO:N/A', 'GeneOrganismRelationship', 'AUTO:N/A')\n",
      "('AUTO:N/A', 'ProteinProteinInteraction', 'AUTO:N/A')\n",
      "('AUTO:N/A', 'ProteinOrganismRelationship', 'AUTO:N/A')\n",
      "('AUTO:N/A', 'GeneMoleculeInteraction', 'AUTO:N/A')\n",
      "('AUTO:N/A', 'ProteinMoleculeInteraction', 'AUTO:N/A')\n",
      "Iteration 2relations:\n",
      "Iteration 3relations:\n",
      "\n",
      "On document PMID10890883_abstract\n",
      "-----------------------------------------------\n",
      "Iteration 1relations:\n",
      "('AUTO:NPR1', 'GeneProteinInteraction', 'AUTO:PR-1')\n",
      "Iteration 2relations:\n",
      "('AUTO:NPR1', 'GeneProteinInteraction', 'AUTO:PR-1')\n",
      "('AUTO:SAR', 'GeneOrganismRelationship', 'NCBITaxon:3702')\n",
      "('GO:0140467', 'GeneOrganismRelationship', 'NCBITaxon:3702')\n",
      "('PR:000011377', 'ProteinProteinInteraction', 'PR:000011377')\n",
      "('AUTO:NPR1', 'ProteinOrganismRelationship', 'NCBITaxon:3702')\n",
      "Iteration 3relations:\n",
      "('AUTO:NPR1', 'GeneProteinInteraction', 'AUTO:PR-1')\n",
      "('AUTO:SAR', 'GeneOrganismRelationship', 'NCBITaxon:3702')\n",
      "('GO:0140467', 'GeneOrganismRelationship', 'NCBITaxon:3702')\n",
      "('PR:000011377', 'ProteinProteinInteraction', 'PR:000011377')\n",
      "('PR:000011377', 'ProteinOrganismRelationship', 'NCBITaxon:3702')\n",
      "\n",
      "On document PMID10476063_abstract\n",
      "-----------------------------------------------\n",
      "Iteration 1relations:\n",
      "Iteration 2relations:\n",
      "('AUTO:not%20found', 'GeneGeneInteraction', 'AUTO:not%20found')\n",
      "('AUTO:not%20found', 'GeneProteinInteraction', 'AUTO:not%20found')\n",
      "('AUTO:not%20found', 'ProteinProteinInteraction', 'AUTO:not%20found')\n",
      "('AUTO:pad3-1', 'GeneMoleculeInteraction', 'CHEBI:22990')\n",
      "('PR:000012221', 'ProteinMoleculeInteraction', 'CHEBI:22990')\n",
      "Iteration 3relations:\n",
      "('AUTO:not%20found', 'GeneGeneInteraction', 'AUTO:not%20found')\n",
      "('AUTO:not%20found', 'GeneOrganismRelationship', 'AUTO:not%20found')\n",
      "('AUTO:not%20found', 'ProteinOrganismRelationship', 'AUTO:not%20found')\n",
      "('AUTO:pad3-1', 'GeneMoleculeInteraction', 'CHEBI:22990')\n",
      "('PR:000012221', 'ProteinMoleculeInteraction', 'CHEBI:22990')\n"
     ]
    }
   ],
   "source": [
    "for doc in evaled_ids:\n",
    "    print('\\nOn document', doc)\n",
    "    print('-----------------------------------------------')\n",
    "    for n, rel_df in {'1': rel_df_1, '2': rel_df_2, '3': rel_df_3}.items():\n",
    "        rel_df = rel_df[rel_df['doc_key'] == doc]\n",
    "        triples = [(row.subject, row.predicate, row.object) for i, row in rel_df.iterrows()]\n",
    "        print('Iteration ' + n + 'relations:')\n",
    "        for trip in triples:\n",
    "            print(trip)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ontogpt",
   "language": "python",
   "name": "ontogpt"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
